[BE-113] 칸반 보드 Columns 조회 시 QueryString으로 year 정보를 받는 V2 API 추가, 지원서 칸반보드 열 생성 V2 API 추가

injae-348 opened the pull request:
### 개요
close #297 

###  작업사항
- `Columns`에 `year(기수)` 필드 추가
- `navigationId`와 `year(기수)` 기반 `Columns`리스트 조회 필드 추가
- `navigationId`와 `year(기수)`를 기반으로 해당하는 모든 세로줄을 조회하는 V2 엔드포인트 추가
- `navigationId`로 모든 세로줄을 조회하는 V1 엔드포인트 `@Deprecated` 설정

- `navigationId`와 `year`로 `List<Columns>`를 가져오며, `CardService.java`에서 Stream을 통해 기수를 필터링 하는 로직 제거
-  `year` 필드가 추가되며, 생성 시 `year`와 `title`을 Body로 받아오는 V2 API 추가

### 변경로직
#### BoardRestController.java
- `@Deprecated` 추가
```java
    @Deprecated
    @Operation(
            summary = "지원서 세로줄 조회(by NavigationId)",
            description = "navigationId에 해당하는 모든 세로줄을 조회합니다., 세로줄이 없으면 빈 배열을 반환합니다.")
    @GetMapping("/boards/navigations/{navigation-id}/columns")
    public ResponseEntity<List<ColumnsResponseDto>> getBoardColumnByNavigationId(
            @PathVariable("navigation-id") Integer navigationId) {
        return new ResponseEntity<>(
                boardLoadUseCase.getColumnsByNavigationId(navigationId), HttpStatus.OK);
    }
```
#### RecordController.java
- `@Deprecated` 위치 변경 - 가독성 향상
```java
    @Deprecated(since = "2024-02-20", forRemoval = true)
    @Operation(summary = "지원자의 면접기록을 전부 조회합니다")
    @ApiErrorExceptionsExample(RecordFindExceptionDocs.class)
    @GetMapping("/records/all")
    public ResponseEntity<List<RecordResponseDto>> findAll() {
        List<Record> records = recordUseCase.findAll();
        return new ResponseEntity<>(RecordResponseDto.from(records), HttpStatus.OK);
    }
```

### 참고사항

- `Columns`에 `year`필드를 추가하였기에 개발 혹은 운영 서버에 적용 된다면, `year(기수)`필드에 default 값으로 28을 주면 좋을 것 같습니다.(2025-3월 기준 29기 모집 중)
- 기존 v1 API는 그대로 둔 채로 새로운 v2 API를 만들었습니다. 개발서버에서 v2 API에 대한 추가 테스트 진행해보고 문제가 없다면, v1과 관련된 로직은 제거하도록 하겠습니다.

injae-348 committed:
[feat]: year필드를 추가하고 관련 도메인 로직 반영

- Columns: year 필드 추가
- BoardLoadUseCase: 조회 시 navigationId와 year를 기준으로 검색하는 로직 추가
- BoardService: BoardLoadUseCase에서 추가한 getColumnsByNavigationIdAndYear 구현

injae-348 committed:
[feat]: Columns 조회 시 year 필터링 기능 추가

- ColumnRepository: findByNavigationIdAndYear 메서드 추가
- ColumnLoadPort: getColumnsByNavigationIdAndYear 메서드 추가
- ColumnAdaptor: getColumnsByNavigationIdAndYear 구현
  - 조회 결과 없는 경우 null 반환

injae-348 committed:
[feat]: 지원서 세로줄 조회 V2 API 추가

- 기존 BoardRestController: navigationId를 기준으로 지원서 세로줄 조회
- 추가 BoardRestControllerV2: navigationId와 year(기수)를 바탕으로 지원서 세로줄 조회

injae-348 committed:
[refactor]: Deprecated 설정 추가 및 위치 수정

- BoardRestController: /api/v1/boards/navigations/{navigation-id}/columns 엔드포인트 @Deprecated 추가
- RecordController: /api/v1/records/all 엔드포인트 @Deprecated 위치 최상단으로 배치해 가독성 향상

rlajm1203 commented:
PR 작성 시, Reviewer 지정 부탁드리겠습니다.

injae-348 commented:
> PR 작성 시, Reviewer 지정 부탁드리겠습니다.

넵! 다음 PR 부터 지정 추가하도록 하겠습니다!

BlackBean99 commented on server/Recruit-Api/src/main/java/com/econovation/recruit/api/card/controller/BoardRestControllerV2.java:
```diff
@@ -0,0 +1,44 @@
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@SecurityRequirement(name = "access-token")
+@RestController
+@RequestMapping("/api/v2")
+@RequiredArgsConstructor
+@Tag(name = "[2.0]. 칸반보드 API V2", description = "칸반보드 관련 API")
+public class BoardRestControllerV2 {
+    private final BoardLoadUseCase boardLoadUseCase;
+
+    @Operation(
+            summary = "지원서 세로줄 조회(by NavigationId and Year)",
+            description = "navigationId와 year(기수)에 해당하는 모든 세로줄을 조회합니다., 세로줄이 없으면 빈 배열을 반환합니다.")
+    @GetMapping("/boards/navigations/{navigation-id}/columns")
+    public ResponseEntity<List<ColumnsResponseDto>> getBoardColumnByNavigationId(
+            @PathVariable("navigation-id") Integer navigationId, @RequestParam("year") Integer year) {
+        return new ResponseEntity<>(
+                boardLoadUseCase.getColumnsByNavigationIdAndYear(navigationId, year), HttpStatus.OK);
+    }
+}
```
Controller class suffix에 V2를 달아둔 이유가 무엇일까요?
이미 릴리즈가 3.0이 넘어갔는데 api version에 대한 명시를 사전에 논의하지는 않았습니다. ( /api/v2 에 대한 설명을 해주실 수 있을까요? )
하위 호환성을 고려한다면 기존 클래스에 메소드를 추가하는게 어땠을까 싶은데요.

해당 테스트를 진행하면서 Spring Security url constraint에 걸리지는 않는지도 검토가 필요해 보입니다. 아마 기존거는 /api/v1을 prefix로 다 걸어놨을 거거든요.

추가로 기수를 param으로 받을 경우라면 앞으로 계속 바뀌는 합격 일자와 현행 지원 기수 이 데이터는 별도로 DB에 보관하는 방법도 고민해보면 좋겠습니다!

injae-348 replied:
Suffix 관련한 부분은 강의에서 컨트롤러의 버전을 관리할 때 V1 부터 V4 Suffix를 사용하는 걸 보고 이전 컨트롤러와 구분을 위해 추가하게 되었습니다.
이 부분에 대해서는 더 나은 네이밍 컨벤션이 있는지 혹은 버전 관리를 할 때는 하나의 클래스 내에서 prefix를 다르게 해 관리를 하는 편인지 여쭤보고 싶었습니다.

/api/v2의 도입의 경우, 기존 API에서 QueryString으로 year 필드를 필수로 받아온다면, 해당 변경 사항은 명세가 변경되는 부분이라 생각해 Github Discussions에서 해당 이야기를 진행하고 추진하고자 했습니다.

해당 API를 구현할 당시 기존 클래스에 메소드를 추가하는 것은 단순 필드 하나만 추가한 API를 만든다면, 한 클래스 내에 중복을 만들게 되어 지양해야한다고 생각했습니다.

`Spring Security url constraint`의 경우 고려하지 못한 부분이었는데 언급해주셔서 감사합니다. /api/v2 prefix 관련 부분은 주의하도록 하겠습니다.

현재 합격 일자와 현행 지원 기수의 경우 환경 변수로 설정해두고 사용하는 걸로 알고 있습니다. `Columns`에 `year(기수)`가 추가된 것처럼 해당 부분 또한 별도의 DB에 보관할 수 있는지 고민해보겠습니다.

BlackBean99 replied:
하나의 클래스 내에서 url마다 별도로 달아줄 순 있는데 
그럼 클래스 단위로 붙여줘야 해서 1개만 달아도 되는걸 이제 메소드 단위로 내려야 합니다.  한 클래스 내에서 2가지 버전이 혼재하는건 SRP 를 준수하지 못한다는 관점으로 볼 수 있겠죠? 만약 v1 을 deprecated해야 된다면 X-Deprecated 응답 헤더로 안내 해주는 것도 방법이 될 수 있겠습니다.
Controller version은 프론트 분들이랑 정말 잘 이야기 해봐야할 중요한 부분입니다. 별개로, 버전 숫자가 올라가는건 저도 찬성입니다. 지속적으로 컨트롤러가 변경되고 개선되고 있다는 증명중 하나니까요 ㅎ. 다만 유지보수 관점에서 비용이 좀 늘어나긴 합니다. (문서화 필수) - 필요하시다면 Swagger에 제가 Reflection으로 커스텀한 코드들이 있는데 거기에 url 파라미터 가져와서 V2에 대한 문서 Description을 추가해주시면 더 프론트 분들이 이해하기 쉬울 것 같아요 (이런 부분을 DX 개선이라고 하죠)

Spring Security Context Holder를 만들어서 UserId 를 가져오고 있는데 UserDetail을 커스텀해서 ContextHolder에 현 모집중 기수인 year을 받아오는 아이디어가 생각났어요.
/Recruit-Api/src/main/java/com/econovation/recruit/api/config/security/SecurityUtils.java 에서 제가 getUserId()를 만들어서 언제든 유저 정보를 가져올 수 있도록 해두었는데요.
만약에 UserDetail을 Custom하면 별도 DB를 운영하지 않더라고 Client Request만으로도 알 수 있다면? 코드가 더 깔끔해지지 않을까요? 별도의 조회 로직이 필요없어지니까요. ( Table 안만들어도 되고? )

저도 year을 .env에 매번 넣어두고 빌드때마다 DI 해주는게 불만이긴 했거든요. 

암튼, @injae-348 인재님 같이 고민해주셔서 감사합니다. 더 좋은 아이디어있으면 언제든 반박해주십쇼 (꾸-벅)

injae-348 replied:
한 클래스 내에서 2가지 버전을 사용하는 것에 대해 위화감이 느껴졌었는데, SRP에 위배되는 관점이였네요.
커스텀 헤더 등록을 통해 해당 API 들에 대해 응답을 추가해보는 방법 알려주셔서 감사합니다!

API Version 관련해서는 FE 분과 잘 이야기 해서 진행하도록 하겠습니다!! 여러 번 강조하시는 만큼 매우 중요하다는 생각이 들었습니다.

문서화 관련해서는 해당 부분 처럼 v1에 있던 부분은 Deprecated 처리하고 칸반보드 API V2 이렇게 적용시켜 보았습니다.
<img width="605" alt="image" src="https://github.com/user-attachments/assets/bd4be659-5c3d-4907-8e86-1c488b16e0c7" />


> 만약에 UserDetail을 Custom하면 별도 DB를 운영하지 않더라고 Client Request만으로도 알 수 있다면? 코드가 더 깔끔해지지 않을까요? 별도의 조회 로직이 필요없어지니까요. ( Table 안만들어도 되고? )

제가 이해한 것으로는 `SecurityContextHolder`에서 가져오는 `UserId`의 경우 Idp 즉 면접관으로 로그인한 사용자의 정보를 가져오는데 여기에 기수(year) 정보를 채워두고 이를 가져와 조회 시 사용한다는 걸까요? 해당 코멘트 부분을 제가 제대로 이해했는지 궁금합니다..!

injae-348 replied:
Spring Security url 제약 조건의 경우 `SecurityConfig.java`에서 확인 하였고 /api/v1/boards/** 와 관련된 조건은 없어 v2 API에 대해 추가 설정 안 해도 될 것 같습니다!

BlackBean99 replied:
SecurityContextHolder에 SecurityUtils라는 클래스에 static 함수로 userId를 유저가 보내주지 않아도 인증정보에서 파싱해서 사용하는 방법을 쓰고 있습니다.
이걸 응용하면 SecurityUtils 클래스에 User가 로그인 할 경우 year를 받아올 수 있도록 getYear() 이런 함수를 만들어볼 수 있겠죠. (Spring Security UserDetail을 커스텀하면 가능할겁니다).  로그인된 유저의 인증 정보와는 상관 없다고 지적할 순 있겠지만, 현 운영중인 기수 정보를 매번 가져올 수 있으면 디폴트로 기수를 안보내줘도 맞춤 정보를 줄 수 있겠죠.

아니면 저장을 따로 하지 않고 현 운영 기수를 계산해보는 방법을 만들어볼 수도 있을 것 같아요.

```java
    public static int getRecruitmentBatch(int year, boolean isFirstHalf) {
        int startYear = 2025; // 2025년 상반기가 29기
        int startBatch = 29;

        int yearDiff = year - startYear;
        int termOffset = isFirstHalf ? 0 : 1; // 상반기면 0, 하반기면 1 추가

        return startBatch + (yearDiff * 2) + termOffset;
    }

```

rlajm1203 replied:
오호 SecurityContext Holder 에서 getYear 메소드를 두고, 현재 모집 기수가 어떤 기수인지 계산해서 가져오면 된다는 말씀이시군요!
그렇게 된다면, 환경변수로 매번 바꿔서 실행해줄 필요도 없으니까요!

BlackBean99 replied:
아니요아니요.

1번방법 : SecurityContextHolder를 Client가 보내주도록 설계해서 꺼내 쓴다
2번방법 : static method로 그냥 year를 계산해서 쓰는 방법

ContextHolder에서는 year를 계산해서는 안되죠.

BlackBean99 commented on server/Recruit-Domain/src/main/java/com/econovation/recruitdomain/domains/board/adaptor/ColumnAdaptor.java:
```diff
@@ -50,6 +50,15 @@ public List<Columns> getColumnsByNavigationId(Integer navigationId) {
+    @Override
+    public List<Columns> getColumnsByNavigationIdAndYear(Integer navigationId, Integer year) {
+        List<Columns> byNavigationIdAndYear = columnRepository.findByNavigationIdAndYear(navigationId, year);
+        if (byNavigationIdAndYear.isEmpty()) {
+            return null;
+        }
```
null return 보다는 위에서 Collections.emptyList를 사용했으니 통일하는건 어떨까요? null 던지는거는 FP 스타일이 아니면 지양하는게 맞다고 생각합니다.

injae-348 replied:
`BoardService`를 확인해보니 `ColumnAdaptor`에서 `null`로 반환된 객체를 체크 후 `Collections.emptyList()`로 반환해주고 있었습니다.
`List`의 특징을 생각해보면 비어있을 때 `null` 보다는 `Collections.emptyList()`를 반환하는게 안정성 있고 일관성을 챙길 수 있을 것 같아 `Collections.emptyList()`를 반환하는 방식으로 수정하겠습니다.

`#BoardService.java`
```java
    @Override
    public List<ColumnsResponseDto> getColumnsByNavigationIdAndYear(Integer navigationId, Integer year) {
        List<Columns> columns = columnLoadPort.getColumnsByNavigationIdAndYear(navigationId, year);
        if (columns == null) {
            return Collections.emptyList();
        }
        return ColumnsResponseDto.from(columns);
    }
```

BlackBean99 replied:
```java
import io.vavr.collection.List;
import io.vavr.control.Option;

@Override
public List<ColumnsResponseDto> getColumnsByNavigationIdAndYear(Integer navigationId, Integer year) {
    return Option.of(columnLoadPort.getColumnsByNavigationIdAndYear(navigationId, year))
                 .map(ColumnsResponseDto::from)
                 .getOrElse(List.empty());
}
```
한창 리크루트 설계할 때 VAVR을 이용한 Flow 를 처리하는 다양한 방법들을 적용해봤는데요
위 코드처럼 if == null 이런 조건을 추가하면 if indent 여러개 늘어나면서 의미를 전달하는데 어려움이 있었습니다.
그래서 FP 스럽게 짜다보니 코드가 가독성이 참 좋아지더라구요 ㅎ  이런 코드들도 참고해보십쇼.
vavr을 안쓰더라도  Optional 을 이용하면 위 코드를
```java
import java.util.Collections;
import java.util.List;
import java.util.Optional;

@Override
public List<ColumnsResponseDto> getColumnsByNavigationIdAndYear(Integer navigationId, Integer year) {
    return Optional.ofNullable(columnLoadPort.getColumnsByNavigationIdAndYear(navigationId, year))
                   .map(ColumnsResponseDto::from)
                   .orElseGet(Collections::emptyList);
}
```
이렇게 비슷하게 짜볼 수 있을겁니다 ㅎㅎ

if문을 다르게 표현하는 다양한 방법을 고민해보시면 분명 큰 도움될겁니다

injae-348 replied:
우선은 `JPARepository`에서 `navigationId`와 `year`를 기준으로 `List<Columns>`를 가져오는 메서드를 통해 구현 해봤고 `null` 이 아닌 빈 리스트가 반환 확인했습니다. `null`과 관련된 로직을 처리한다면, VAVR, Optional로  `null`일 가능성을 포함해 indent를 줄이는 부분 참고하도록 하겠습니다! 리뷰 감사합니다!
```java
    @Override
    public List<ColumnsResponseDto> getColumnsByNavigationIdAndYear(Integer navigationId, Integer year) {
        List<Columns> columns = columnLoadPort.getColumnsByNavigationIdAndYear(navigationId, year);
        return ColumnsResponseDto.from(columns);
    }
```

injae-348 committed:
[refactor]: navigationId로 조회에서 navigationId와 year로 Columns 리스트 반환

- ColumnService: getByNavigationId에서 getByNavigationIdAndYear로 변경
- ColumnsUseCase: getByNavigationId에서 getByNavigationIdAndYear로 변경

injae-348 committed:
[refactor]: getByNavigationId 내부 로직 수정

- CardService: Columns 필터링 수정, 예외처리 추가
  - getByNavigationId 대신 getByNavigationIdAndYear 적용
    기수 필터링을 위한 스트림 순환 제거
  - Columns에 해당 기수(year)가 없는 경우 커스텀 예외 반환

injae-348 committed:
[feat]: Columns 생성 시 title과 year를 받아오는 DTO 추가

injae-348 committed:
[refactor]: findByNavigationIdAndYear함수 @Query 추가

injae-348 committed:
[feat]: Columns 생성 시 Year 필드를 받아오는 메서드 추가 및 기존 메서드에 환경변수로 year 필드 적용

- BoardRegisterUseCase: createColumnWithYear 추가
- BoardService: createColumnWithYear 구현 및 기존 createColumn 메서드에
  환경변수로 year 필드 자동 할당

injae-348 committed:
[feat]: 지원서 칸반보드 열 생성 V2 API 추가

- BoardRestControllerV2: title과 year를 Body로 받아와 Column을 생성하는
  엔드포인트 추가

injae-348 committed:
[refactor]: 지원서 칸반보드 열 생성 v1 API @Deprecated 추가 및 since 적용

- BoardRestController
  - 지원서 칸반보드 열(세로줄) 생성 v1 API @Deprecated 추가
  - 지원서 세로줄 조회(by NavigationId) v1 API @Deprecated에 since 옵션 추가

injae-348 committed:
[refactor]: List 반환 값으로 null 대신 EmptyList 반환

- ColumnAdaptor: navigationId와 year로 Columns List를 가져올 때 null 대신 EmptyList 반환
- BoardService: getColumnsByNavigationIdAndYear 결과에 대해 null 검사 제거

injae-348 committed:
[style]: spotless 적용

injae-348 committed:
[style]: spotless 적용

injae-348 committed:
[refactor]: columns null 검사 if 문에서 Optional로 변경

Profile-exe commented on server/Recruit-Domain/src/main/java/com/econovation/recruitdomain/domains/board/domain/ColumnRepository.java:
```diff
@@ -3,12 +3,16 @@
 public interface ColumnRepository extends JpaRepository<Columns, Integer> {
     List<Columns> findByNavigationId(Integer navigationId);
 
+    @Query("SELECT c FROM Columns c WHERE c.navigationId = :navigationId AND c.year = :year")
```
JPQL을 사용한 이유가 궁금합니다!

Profile-exe reviewed (commented):
수고하셨습니다!

rlajm1203 committed:
Merge branch 'develop' into feature/BE-113
